lib/Serialiser/Ntriples.php

```diff
@@ -49,27 +49,143 @@
+     *
+     * The sequence with which the characters are replaced is important.
+     *
+     * The sequence of replacements is:
+     * - Backslash (\) - Escape character. It is escaped first so that we do not
+     *  double escape other sequences that contain a backslash.
+     * - Control characters (0-31) and DEL (127) - except \t, \n, \r and \".
+     *  These are replaced with their unicode representation using a normal
+     *  str_replace because they break preg_replace_callback.
+     * - 8-byte characters - These are replaced with their unicode
+     * representation using a preg_replace_callback. They are replaced first so
+     * that they are not replaced as multiple characters down below.
+     * - 4-byte characters - These are replaced with their unicode
+     * representation using a preg_replace_callback. They are replaced next so
+     * that they are not replaced as multiple characters down below.
+     * - Characters from 127 to 255 - These are replaced with their unicode
+     * representation using a normal str_replace because fail to match in
+     * preg_replace_callback. This takes place here because otherwise, they
+     * might be confused with multibyte characters if it takes place before the
+     * two steps above. And it takes place after the above two steps, in order
+     * to avoid replacing them as multiple characters.
+     * - Characters from 32 to 126 - These are not escaped as they are printable
+     * characters. However, the double quote (\") is escaped above as it is a
+     * special character in N-Triples.
+     *
+     * @see https://www.w3.org/TR/n-triples/#n-triples-grammar
+     *
+     * @param string $str
+     *   The string to escape.
+     *
+     * @return string
+     *   The escaped string.
      */
     protected function escapeString($str)
     {
-        $result = '';
-        $strLen = mb_strlen($str, 'UTF-8');
+        // Escape the "\" character so that we do not double escape it.
+        $str = str_replace('\\', '\\\\', $str);
+        // List of characters indexed by their printed representation.
+        $special_control_chrs = [];
+        foreach (range(0, 31) as $i) {
+            $special_control_chrs[chr($i)] = $this->unicodeChar($i);
+        }
 
-        for ($i = 0; $i < $strLen; ++$i) {
-            $c = mb_substr($str, $i, 1, 'UTF-8');
+        // However, "\t", "\n", "\r" and "\"" are allowed.
+        $special_control_chrs[chr(9)] = '\t';
+        $special_control_chrs[chr(10)] = '\n';
+        $special_control_chrs[chr(13)] = '\r';
+        $special_control_chrs[chr(34)] = '\\"';
 
-            if (!isset($this->escChars[$c])) {
-                $this->escChars[$c] = $this->escapedChar($c);
-            }
+        // Handle also the DEL character.
+        $special_control_chrs[chr(127)] = $this->unicodeChar(127);
+
+        $str = str_replace(array_keys($special_control_chrs), array_values($special_control_chrs), $str);
```

Comment:
I think it's safe to run a single `str_replace()` (i.e. add backslash escaping as 1st replacements array element). If read correctly the docs at https://www.php.net/manual/en/function.str-replace.php:

> If search or replace are arrays, their elements are processed first to last.